libs/server/Objects/Set/SetObjectImpl.cs

```diff
@@ -256,5 +258,107 @@ private void SetPop(byte* input, int length, ref SpanByteAndMemory output)
+                    // Return an array of distinct elements
+                    var countParameter = count > set.Count ? set.Count : count;
+                    List<byte[]> randomElements = new List<byte[]>();
+
+                    // Write the size of the array reply
+                    while (!RespWriteUtils.WriteArrayLength(countParameter, ref curr, end))
+                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+
+                    while (randomElements.Count < countParameter)
+                    {
+                        int index = RandomNumberGenerator.GetInt32(0, set.Count);
+                        var item = set.ElementAt(index);
+                        if (!randomElements.Any(x => x.SequenceEqual(item)))
+                        {
+                            randomElements.Add(item);
+                            while (!RespWriteUtils.WriteBulkString(item, ref curr, end))
+                                ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                        }
+                        countDone++;
+                    }
+                    countDone += count - countParameter;
+                }
+                else if (count == int.MinValue) // no count parameter is present
+                {
+                    // Return a single random element from the set
+                    if (set.Count > 0)
+                    {
+                        int index = RandomNumberGenerator.GetInt32(0, set.Count);
+                        var item = set.ElementAt(index);
+                        while (!RespWriteUtils.WriteBulkString(item, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    }
+                    else
+                    {
+                        // If set is empty, return nil
+                        while (!RespWriteUtils.WriteDirect(CmdStrings.RESP_ERRNOTFOUND, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    }
+                    countDone++;
+                }
+                else
+                {
+                    // Return an array with potentially duplicate elements
+                    int countParameter = Math.Abs(count);
+
+                    // Write the size of the array reply
+                    while (!RespWriteUtils.WriteArrayLength(countParameter, ref curr, end))
+                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    for (int i = 0; i < countParameter; i++)
+                    {
+                        if (set.Count > 0)
+                        {
+                            int index = RandomNumberGenerator.GetInt32(0, set.Count);
```

Comment:
Again, a bit inefficient, since you're iterating over the set n times. If you generate the indexes first, you can do it with one pass over the set :)